homeassistant/auth/providers/command_line.py

```diff
@@ -0,0 +1,149 @@
+"""Auth provider that validates credentials via an external script."""
+
+import typing as T
+
+import asyncio.subprocess
+import collections
+import logging
+
+import voluptuous as vol
+
+from homeassistant.exceptions import HomeAssistantError
+
+from . import AuthProvider, AUTH_PROVIDER_SCHEMA, AUTH_PROVIDERS, LoginFlow
+from ..models import Credentials, UserMeta
+
+
+CONF_PROGRAM = "program"
+CONF_ARGS = "args"
+
+CONFIG_SCHEMA = AUTH_PROVIDER_SCHEMA.extend({
+    vol.Required(CONF_PROGRAM): str,
+    vol.Optional(CONF_ARGS, default=None): vol.Any(vol.DefaultTo(list), [str]),
+}, extra=vol.PREVENT_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class InvalidAuthError(HomeAssistantError):
+    """Raised when authentication with given credentials fails."""
+
+
+@AUTH_PROVIDERS.register("command_line")
+class CommandLineAuthProvider(AuthProvider):
+    """Auth provider validating credentials by calling a command."""
+
+    DEFAULT_TITLE = "Command Line Authentication"
+
+    # which keys to accept from a program's stdout
+    ALLOWED_META_KEYS = ("name",)
+
+    def __init__(self, *args: T.Any, **kwargs: T.Any) -> None:
+        """Extend parent's __init__.
+
+        Adds self._user_meta dictionary to hold the user-specific
+        attributes provided by external programs.
+        """
+        super().__init__(*args, **kwargs)
+        self._user_meta = {}  # type: T.Dict[str, T.Dict[str, T.Any]]
+
+    async def async_login_flow(self, context: T.Optional[T.Dict]) -> LoginFlow:
+        """Return a flow to login."""
+        return CommandLineLoginFlow(self)
+
+    async def async_validate_login(self, username: str, password: str) -> None:
+        """Validate a username and password."""
+        env = {
+            "username": username,
+            "password": password,
+        }
+        try:
+            process = await asyncio.subprocess.create_subprocess_exec(
+                self.config[CONF_PROGRAM], *self.config[CONF_ARGS],
+                env=env,
+                stdout=asyncio.subprocess.PIPE,
+            )
+            stdout = (await process.communicate())[0]
+        except OSError as err:
+            _LOGGER.error("Error while authenticating '%s': %s",
+                          username, err)
+            raise InvalidAuthError
+
+        if process.returncode != 0:
+            raise InvalidAuthError
+
+        meta = {}  # type: T.Dict[str, str]
+        for _line in stdout.splitlines():
+            try:
+                line = _line.decode().lstrip()
+                if line.startswith("#"):
+                    continue
+                key, value = line.split("=", 1)
```

Comment:
This way to grab metadata is too specific for your usage, I would suggest generalize it by using template(possible?) or regex match

Reply:
What kind of metadata do you think this is not suited for? The only thing that's allowed to be set by HA right now is the name, and there will probably be groups in the future, but what else?

Reply:
I think that templates and regex are a bit too error prone, and this is sensitive matter once it includes groups. So I suggest that we just define a format, and I think in that case the KEY=VALUE format is ok? Especially if it's a bash script, it might not be easy to define a more complicated output.

Reply:
Yes, exactly. And instead of having to maintain regexes in the HA configuration and trust they're applied as I expect, I feel much more comfortable with adding the formatting directly to the adapter that integrates my auth backend. I think no one will use something like ``ldapsearch`` directly without a script around.

Maybe we could make it configurable whether to read meta variables from stdout or not, in case somebody really wants to use something he can't control the output of. What about a ``meta: false`` option which you have to turn on explicitly? Could easily be described in the docs.

@balloob What do you mean with MVP?

Reply:
Agreed on ``meta`` parameter and current implementation as discussed below. Marking this resolved.